File: client/objects.js
Wayfarer.world_object = function () {
	// position related infos
	this.x = 0;
	this.y = 0;
	this.heading=180;
	
	// drawing related infos
	this.width=20;
	this.height=20;
	
	// datastore related infos
	this.remote_id = null;
	this.last_updated_at = null;
	
	this.update = function(data) {
		if (data['x']) this.x = data['x'];
		if (data['y']) this.y = data['y'];
		if (data['heading']) this.heading = data['heading'];
	}
	
	// We're doing this ourselves because the toJSON needs to be *lightning* fast
	this.toJSON=function() {
		return '{"game_id":'+this.remote_id+',"x":'+this.x+', "y":'+this.y+',"heading":'+this.heading+'}';
	}
	
	this.draw=function(canvas) {
		// This is where we'll put our own drawing code.
		// This is some default drawing code
		canvas.save();
		canvas.translate(this.x,this.y);
		canvas.rotate((-this.heading+180) * Math.PI / 180);
		canvas.drawImage(Wayfarer.default_image,-(this.width/2),-(this.height/2),this.width,this.height);
		canvas.restore();
	}
	
	// User-specified stuff for your own game objects
	this.tick=function() {
		// This is where we put things that happen every 'tick'
		// Good examples are friction functions, position updates for player objects, etc. etc.
	}
	
	this.click = function() {};
	this.selected = function() {};
	this.deselected = function() {};
	this.collided = function(obj) {};
	this.keyDown = function(code) {};
	this.keyUp = function(code) {};
	
	this.initialized = function() {};
	this.destroyed = function() {};
};

Wayfarer.moveable_object = function() {
	this.inheritFrom = Wayfarer.world_object;
	this.inheritFrom();
	
	this.v = 0;
	
	this.tick=function() {
		this.x += Math.sin(this.heading * Math.PI / 180)*this.v;
		this.y += Math.cos(this.heading * Math.PI / 180)*this.v;
	};
};

Wayfarer.player_object = function () {
	this.inheritFrom=Wayfarer.moveable_object;
	this.inheritFrom();
	
	this.rotation_speed=0; // 1 = rotating CW, -1 = rotating CCW, 0 = not rotating
	this.acceleration=0;
	
	this.v_prime_speed = 0.1; // pixels per tick per tick
	this.max_v = 2;
	
	this.rotation_speed = 0; // degrees per tick per tick
	
	this.tick = function() {
		this.x += Math.sin(this.heading * Math.PI / 180)*this.v;
		this.y += Math.cos(this.heading * Math.PI / 180)*this.v;
		
		// Input related stuff
		
		if (1==1) { // TODO: packet dropping
			this.heading+=this.rotation_speed;
			this.v+=this.acceleration;
			if (this.v>this.max_v) this.v=this.max_v;
			if (this.v>0) this.v-=(this.v / 10); // friction IN SPAAAACCEEEEE
		}
		
	}
	
	this.toJSON=function() {
		return '{"game_id":'+this.remote_id+',"rotation_speed":'+this.rotation_speed+',"acceleration":'+this.acceleration+'}';
	}
	
	this.keyDown = function(code) {
		//Wayfarer.log(code);
		switch(code) {
			case 37: /* left-arrow */
				if (this.rotation_speed != 1) {
					this.rotation_speed = 1;
					Wayfarer.Network.request('input',{object:this.toJSON()});
				}
			break;
			case 38: /* up-arrow */
				if (this.acceleration < this.v_prime_speed) {
					this.acceleration=this.v_prime_speed;
					Wayfarer.Network.request('input',{object:this.toJSON()});
				}
			break;
			case 39: /* right-arrow */
				if (this.rotation_speed != -1) {
					this.rotation_speed = -1;
					Wayfarer.Network.request('input',{object:this.toJSON()});
				}
			break;
		};
	};
	
	this.keyUp = function(code) {
		switch(code) {
			case 37: /* left-arrow */
			case 39: /* right-arrow */
				if (this.rotation_speed != 0) {
					this.rotation_speed = 0;
					Wayfarer.Network.request('input',{object:this.toJSON()});
				}
			break;
			case 38: /* up-arrow */
				if (this.acceleration > 0) {
					this.acceleration = 0;
					Wayfarer.Network.request('input',{object:this.toJSON()});
				}
			break;
		};
	};
}

File: server/objects.rb
module Wayfarer

	module Objects

		class Base
			attr_accessor(:x,:y,:heading, :game_id)
			
			def initialize
				@x=0;@y=0;@heading=0;@game_id = Wayfarer::World.unique_id
			end
		
			def tick
		  end
		
			def to_s
				"#{self.class.to_s} (##{game_id}) - #{x},#{y},#{heading}"
			end
			
			def to_json
				'{"game_id":'+game_id.to_s+',"x":'+x.to_s+',"y":'+y.to_s+',"heading":'+heading.to_s+'}';
			end
		
		end
	
		class Moveable < Base
			attr_accessor(:v_rot,:v)

	    def tick
	    end
		
		end	

		class Player < Base
			attr_accessor(:acceleration, :rotation_speed, :max_rotation, :max_velocity, :ip, :login)
		
			def initialize
				super
				
				@x=rand(100)-50
				@y=rand(100)-50
				@heading = rand(360)
				
				@max_rotation = 1

				@rotation_speed = 0
				@acceleration_speed = 0.1
				@max_velocity = 2.0
				
				@v = 0; @v_rot = 0; @acceleration = 0.0
			end
		
			def tick
				@x += Math.sin(heading * Math::PI / 180)*@v
				@y += Math.cos(heading * Math::PI / 188)*@v
				@heading += @rotation_speed

				# input 
				@v+=@acceleration
				@v=@max_velocity if (@v>@max_velocity)
				@v-=(@v / 10.0) if (@v>0)
			end
	
			def to_s
				"#{self.class.to_s} (##{game_id}) - #{x},#{y},#{heading},#{@v},#{acceleration},#{rotation_speed}"
			end
		
			def to_json
				'{"type":"player","game_id":'+game_id.to_s+',"x":'+x.to_s+
				',"y":'+y.to_s+',"heading":'+heading.to_s+',"player_name":"'+login.to_s+'"}';
			end
		
	  end
	
	end
	
end